source/Halibut.Tests/Support/BackwardsCompatibility/HalibutTestBinaryRunner.cs

```diff
@@ -111,80 +111,94 @@ public async Task<RunningOldHalibutBinary> Run()
             }
 
             settings.Add("ServiceConnectionType", serviceConnectionType.ToString());
+            
+            var tmpDirectory = new TmpDirectory();
 
-            var cts = new CancellationTokenSource();
+            var (task, serviceListenPort, runningTentacleCancellationTokenSource) = await StartHalibutTestBinary(version, settings, tmpDirectory);
 
-            try
-            {
-                var tmp = new TmpDirectory();
-
-                var (task, serviceListenPort) = await StartHalibutTestBinary(version, settings, tmp, cts.Token);
-
-                return new RunningOldHalibutBinary(cts, task, tmp, serviceListenPort, compatBinaryStayAlive);
-            }
-            catch (Exception)
-            {
-                cts.Cancel();
-                cts.Dispose();
-                throw;
-            }
+            return new RunningOldHalibutBinary(runningTentacleCancellationTokenSource, task, tmpDirectory, serviceListenPort, compatBinaryStayAlive);
         }
 
-        async Task<(Task, int?)> StartHalibutTestBinary(string version, Dictionary<string, string> settings, TmpDirectory tmp, CancellationToken cancellationToken)
+        async Task<(Task RunningTentacleTask, int? ServiceListenPort, CancellationTokenSource RunningTentacleCancellationTokenSource)> StartHalibutTestBinary(
```

Comment:
I've refactored the way cancellation works so it's a bit more aggressive to try and clean things up. I doubt this is the underlying cause of hangs but just incase